lib/tcpip_stack_direct.ml

```diff
@@ -77,10 +77,10 @@ struct
   let ipv4 { ipv4; _ } = ipv4
 
   let listen_udpv4 t ~port callback =
-    Hashtbl.replace t.udpv4_listeners port callback
+    Hashtbl.replace t.udpv4_listeners (port mod 65536) callback
```

Comment:
Probably using `abs port` instead of `port` would be better to avoid negative port numbers.  And probably when you get zero, the port should be chosen randomly among the set of free ports.